### PR TITLE
feat(taiko-client): introduce zk

### PR DESCRIPTION
**Pull request by YoGhurt111**

### CODE REVIEW

**Review by smtmfft — APPROVED** [review]

File: .github/workflows/taiko-client--docker.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ name: "Push docker image to GAR"
 
 on:
   push:
-    branches: [main]
+    branches: [feature/introduce-zk2]
```
Comment (mask-pp) [inline .github/workflows/taiko-client--docker.yml:5]:
```suggestion
    branches: [main]
```

File: packages/taiko-client/prover/proof_producer/sgx_and_zkvm_producer.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"time"
+
+	"github.com/taikoxyz/taiko-mono/packages/taiko-client/bindings/metadata"
```
Comment (mask-pp) [inline packages/taiko-client/prover/proof_producer/sgx_and_zkvm_producer.go:15]: order this one.

File: packages/protocol/contracts/L1/tiers/DevnetTierProvider.sol
Comment (YoGhurt111) [inline packages/protocol/contracts/L1/tiers/DevnetTierProvider.sol:1]: Shall we revert this change?

Comment (davidtaikocha) [reply packages/protocol/contracts/L1/tiers/DevnetTierProvider.sol:1]: Lets revert all non-client changes here, and open new PRs if necessary.

File: packages/taiko-client/prover/proof_submitter/proof_submitter.go [hunk 3]
```diff
@@ -158,6 +161,12 @@ func (s *ProofSubmitter) RequestProof(ctx context.Context, meta metadata.TaikoBl
 				header,
 			)
 			if err != nil {
+				// If request proof has timed out in retry, let's cancel the proof generating and skip
+				if errors.Is(err, proofProducer.ErrProofInProgress) && time.Since(startTime) >= ProofTimeout {
+					log.Error("Request proof has timed out, start to cancel", "blockID", opts.BlockID)
+					_ = s.proofProducer.RequestCancel(ctx, opts)
```
Comment (davidtaikocha) [inline packages/taiko-client/prover/proof_submitter/proof_submitter.go:167]: No need to check error here?

Comment (YoGhurt111) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:167]: As said before, it's just a one call here, otherwise we're stuck with complicated retry logic.
Or you think there's a strong need here for error and retry process?

**Review by mask-pp — APPROVED** [review]

File: packages/protocol/contracts/L1/tiers/DevnetTierProvider.sol
Comment (YoGhurt111) [reply packages/protocol/contracts/L1/tiers/DevnetTierProvider.sol:1]: OK, will do

File: packages/taiko-client/prover/proof_submitter/proof_submitter.go (hunk 3 above)
Comment (davidtaikocha) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:167]: If so, lets `log.Error` those errors, completely ignore them is not a good practice.

Comment (YoGhurt111) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:167]: OK

File: packages/protocol/script/DeployOnL1.s.sol [hunk 4]
```diff
@@ -22,13 +22,13 @@ import "../contracts/automata-attestation/AutomataDcapV3Attestation.sol";
 import "../contracts/automata-attestation/utils/SigVerifyLib.sol";
 import "../contracts/automata-attestation/lib/PEMCertChainLib.sol";
 import "../contracts/mainnet/MainnetSgxVerifier.sol";
+import "../contracts/thirdparty/risczero/groth16/RiscZeroGroth16Verifier.sol";
```
Comment (davidtaikocha) [inline packages/protocol/script/DeployOnL1.s.sol:25]: Are these changes expected in this PR?

File: packages/taiko-client/prover/proof_submitter/proof_submitter.go [hunk 5]
```diff
@@ -158,6 +161,13 @@ func (s *ProofSubmitter) RequestProof(ctx context.Context, meta metadata.TaikoBl
 				header,
 			)
 			if err != nil {
+				// If request proof has timed out in retry, let's cancel the proof generating and skip
+				if errors.Is(err, proofProducer.ErrProofInProgress) && time.Since(startTime) >= ProofTimeout {
+					log.Error("Request proof has timed out, start to cancel", "blockID", opts.BlockID)
+					cancelErr := s.proofProducer.RequestCancel(ctx, opts)
+					log.Error("Failed to request cancellation of proof", "err", cancelErr)
```
Comment (davidtaikocha) [inline packages/taiko-client/prover/proof_submitter/proof_submitter.go:168]: check `If cancelErr != nil` at first?

Comment (YoGhurt111) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:168]: Sure

File: packages/protocol/script/DeployOnL1.s.sol (hunk 4 above)
Comment (YoGhurt111) [reply packages/protocol/script/DeployOnL1.s.sol]: Changed location

File: packages/protocol/script/DeployOnL1.s.sol [hunk 6]
```diff
@@ -27,9 +27,9 @@ import "../test/common/erc20/FreeMintERC20.sol";
 import "../test/common/erc20/MayFailFreeMintERC20.sol";
 import "../test/L1/TestTierProvider.sol";
 import "../test/DeployCapability.sol";
+import "../contracts/verifiers/RiscZeroVerifier.sol";
```
Comment (davidtaikocha) [inline packages/protocol/script/DeployOnL1.s.sol:30]: Still need some changes?

Comment (YoGhurt111) [reply packages/protocol/script/DeployOnL1.s.sol]: Yes, did it.

**Review by davidtaikocha — APPROVED** [review]
